File: 560-subarray-sum-equals-k/subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int count = 0;
        int sum = 0;
        Map<Integer, Integer> sumOccurrences = new HashMap<>();
        sumOccurrences.put(0, 1);

        for (int num : nums) {
            sum += num;
            count += sumOccurrences.getOrDefault(sum - k, 0);
            sumOccurrences.put(sum, sumOccurrences.getOrDefault(sum, 0) + 1);
        }

        return count;
    }
}

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        res = 0
        freq = {0:1}
        ps = 0

        for i in range(len(nums)):
            ps += nums[i]
            sub = ps - k
            if sub in freq:
                res += freq[sub]
            freq[ps] = freq.get(ps, 0) + 1

        return res
